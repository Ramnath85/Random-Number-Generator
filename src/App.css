.App-header {
    background-color: #fbfdff;
    display: flex;
    flex-direction: column;
    align-items: center;
    font-size: calc(10px + 2vmin);
    color: #131883;
    flex: 1 1 auto;
    justify-content: center;
}
.App-header h1 {
    font-size: 120px;
    font-weight: 500;
    margin-bottom: 70px;
}

.App-header h3 {
    font-weight: 300;
}

.App-header .sort-details {
    font-size: 20px;
    font-weight: 300;
}
.wrapper {
    flex: 1 1 50%;
    padding: 40px 20px;
}
.App-subtitle {
    display: flex;
    flex-direction: column;
    align-items: center;
    font-size: calc(10px + 2vmin);
}

.App-body {
    display: flex;
    flex-direction: column;
    align-items: center;
}

.value-input {
    margin-left: 10px;
}

.generated-phone-numbers {
    display: flex;
    flex-direction: column;
    align-items: center;
    list-style: none;
    color: #131883;
}

.App-link {
    color: #61dafb;
}

.input-wrapper {
    display: flex;
    justify-content: space-evenly;
    width: 100%;
    padding: 24px;
}

.input-wrapper input[type=number] {
    height: 45px;
    width: 200px;
}
.input-wrapper button {
    padding: 13px;
    font-size: 14px;
    background-color: #131883;
    border-radius: 5px;
    border: 2px solid #fff;
    color: #fff;
    outline: none;
    cursor: pointer;
}

.sortBy select {
    height: 40px;
    width: 200px;
    margin-left: 10px;
}