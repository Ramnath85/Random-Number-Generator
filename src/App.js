import React, { Component, Fragment } from 'react';
import saveAs from 'file-saver';
import './App.css';
import Header from './components/Header/Header';
import SubHeader from './components/Header/SubHeader'
import Input from './components/Input/Input';
import PhoneNumbers from './components/PhoneNumbers/PhoneNumber';
import SortBy from './components/Sort/SortBy';
import Error from './components/Error/Error';
import ExportButton from './components/Export/ExportButton';

/**
 * App class
 */
class App extends Component {
  state = {
    error: false,
    message: "",
    bound: 1,
    phoneNumbers: [],
    max: null,
    min: null,
    total: 0,
    defaultSort: 'asc'
  };

  /**
   * Handle Input
   *
   * @return {object} event
   * @param {object} event
   */
  handleInput = async event => {
    event.preventDefault();
    const bound = event.target.value;
    try {
      if (Math.floor(Number(bound))) {
        this.setState({
          bound
        })
      }
    } catch (e) {
      this.setState({
        error: true,
        message: "An error occurred during this operation"
      })
    }
  };
  /**
   * Handle MinMax
   *
   * @return {object} event
   * @param {object} event
   */
  handleMinMax = () => {
    const { phoneNumbers } = this.state;
    if (phoneNumbers.length > 0) {
      const min = '0' + Math.min(...phoneNumbers);
      const max = '0' + Math.max(...phoneNumbers);
      const total = phoneNumbers.length;
      this.setState({
        min,
        max,
        total
      })
    }
  };
  /**
   * Sort Phone numbers
   */
  sortPhoneNumbers = () => {
    const { defaultSort, phoneNumbers } = this.state;
    if (!phoneNumbers.length > 0) return;
    if(defaultSort === 'asc'){
      this.setState({
        phoneNumbers : phoneNumbers.sort((a, b) => 0 - (a > b ? -1 : 1))
      });
    } else {
      this.setState({
        phoneNumbers : phoneNumbers.sort((a, b) => 0 - (a > b ? 1 : -1))
      })
    }
  };

  /**
   * Handle sort change
   *
   * @return {object} event
   * @param {object} event
   */
  handleSortChange = event => {
    event.preventDefault();
    const sorter = event.target.value;
    this.setState({
      sorter
    }, () => this.sortPhoneNumbers());
  };

  /**
   * Handle number generation
   *
   * @return {object} event
   * @param {object} event
   */
  handleNumberGeneration = event => {
    event.preventDefault();
    const { bound } = this.state;
    if (bound > 10000) return this.setState({
      error: true,
      message: "Please enter a number less than 10000"
    });
    let phoneNumbers = [];
    let phoneNumber = 0;
    while (phoneNumber < bound) {
      phoneNumbers.push('0' + Math.floor(Math.random() * 900000000 + 100000000));
      phoneNumber++;
    }
    return this.setState({
      phoneNumbers
    }, () =>
      this.handleMinMax(),
      this.sortPhoneNumbers()
    );
  };
  /**
   * Export phone numbers
   */
  exportPhoneNumbers = () => {
    const { phoneNumbers } = this.state;
    if (phoneNumbers.length > 0) {
      saveAs(new Blob(phoneNumbers, {
        type: "text/csv;charset=utf-8"
      }), 'generated_phone_numbers.csv')
    }
  };

  render() {
    const { error, message, phoneNumbers, max, min, total } = this.state;
    return (
      <Fragment>
        <Header
          phoneNumbers={phoneNumbers}
          min={min}
          max={max}
          total={total}
        />
        <div className="wrapper">
          <SubHeader />
          <div className="App-body">
            <Error
              error={error}
              message={message}
            />
            <div className="input-wrapper">
              <Input
                onClick={this.handleNumberGeneration}
                onChange={this.handleInput}
              />
              <SortBy
                phoneNumbers={phoneNumbers}
                onChange={this.handleSortChange}
              />
            </div>
            <ExportButton
              phoneNumbers={phoneNumbers}
              onClick={this.exportPhoneNumbers}
            />
            <PhoneNumbers
              phoneNumbers={phoneNumbers}
            />
          </div>
        </div>
      </Fragment>
    );
  }
}

export default App;
