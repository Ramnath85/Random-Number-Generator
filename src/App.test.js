import React from 'react';
import ReactDOM from 'react-dom';
import { configure, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import App from './App';

configure({ adapter: new Adapter() });

describe('App Test', () => {
  let instance;
  let shallowComponent;

  beforeEach(() => {
    shallowComponent = shallow(<App />);
    instance = shallowComponent.instance();
  });

  test('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<App />, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  test('should render the App component', () => {
    expect(shallowComponent).toMatchSnapshot();
    expect(shallowComponent.length).toEqual(1);
  });

  test('should set error to true when bound is greater than 10000 and display a message', () => {
    const event = {
      preventDefault: () => {},
    };

    shallowComponent.setState({ bound: 20000 });
    instance.handleNumberGeneration(event);
    const state = shallowComponent.state();

    expect(state.bound).toEqual(20000);
    expect(state.error).toBe(true);
    expect(state.message).toBe("Please enter a number less than 10000");
  });

  test('should generate phone numbers', () => {
    const event = {
      preventDefault: () => {},
    };
    shallowComponent.setState({ bound: 10 });
    instance.handleNumberGeneration(event);
    const state = shallowComponent.state();

    expect(state.bound).toEqual(10);
    expect(state.error).toBe(false);
  });

  test('should update the details', () => {
    const event = {
      preventDefault: () => {},
    };

    shallowComponent.setState({ bound: 20 });
    instance.handleNumberGeneration(event);
    const state = shallowComponent.state();

    expect(state.total).toBe(20);
    expect(state.min).toBeDefined();
    expect(state.max).toBeDefined();
  });

  test('should set bound', async () => {
    const event = {
      preventDefault: () => {},
      target: {
        value: 10
      }
    };

    await instance.handleInput(event);
    const state = shallowComponent.state();
    expect(state.bound).toBe(10);
  });

  test('should not change the bound', async () => {
    const event = {
      preventDefault: () => {},
      target: {
        value: 'hmmm'
      }
    };
    await instance.handleInput(event);
    const state = shallowComponent.state();
    expect(state.bound).toBe(1);
  });

  test('should set details', () => {
    shallowComponent.setState({ phoneNumbers: ['0903428765', '0787665456']});
    instance.handleMinMax();
    const state = shallowComponent.state();

    expect(state.total).toBe(2);
    expect(state.min).toBeDefined();
    expect(state.max).toBeDefined();
  });

  test('should test sortPhoneNumbers', () => {
    shallowComponent.setState({ defaultSort: 'desc', phoneNumbers: ['0903428765', '0787665456'] });
    instance.sortPhoneNumbers();
    const state = shallowComponent.state();

    expect(state.phoneNumbers.length).toBeGreaterThan(1);
    expect(state.defaultSort).toBe('desc');
  });

  test('should call sortPhoneNumbers', () => {
    shallowComponent.setState({ phoneNumbers: ['0903428765', '0787665456']});
    const event = {
      preventDefault: () => {},
      target: {
        value: 'desc'
      }
    };

    let spy;
    spy = jest.spyOn(instance, 'handleSortChange');
    instance.handleSortChange(event);
    const state = shallowComponent.state();

    expect(state.defaultSort).toBe('asc');
    expect(spy).toHaveBeenCalled();
    spy.mockRestore();
  });

  test('should call sortPhoneNumbers with ascending order', () => {
    shallowComponent.setState({ phoneNumbers: ['0903428765', '0787665456']});
    const event = {
      preventDefault: () => {},
      target: {
        value: 'asc'
      }
    };
    let spy;
    spy = jest.spyOn(instance, 'handleSortChange');
    instance.handleSortChange(event);
    const state = shallowComponent.state();

    expect(state.defaultSort).toBe('asc');
    expect(spy).toHaveBeenCalled();
    spy.mockRestore();
  });
});
