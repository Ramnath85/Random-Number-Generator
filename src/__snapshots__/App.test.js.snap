// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Test should render the App component 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <App />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <Header
          max={null}
          min={null}
          phoneNumbers={Array []}
          total={0}
        />,
        <div
          className="wrapper"
        >
          <SubHeader />
          <div
            className="App-body"
          >
            <Error
              error={false}
              message=""
            />
            <div
              className="input-wrapper"
            >
              <Input
                onChange={[Function]}
                onClick={[Function]}
              />
              <SortBy
                onChange={[Function]}
                phoneNumbers={Array []}
              />
            </div>
            <ExportButton
              onClick={[Function]}
              phoneNumbers={Array []}
            />
            <GeneratedNumbers
              phoneNumbers={Array []}
            />
          </div>
        </div>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "max": null,
          "min": null,
          "phoneNumbers": Array [],
          "total": 0,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <SubHeader />,
            <div
              className="App-body"
            >
              <Error
                error={false}
                message=""
              />
              <div
                className="input-wrapper"
              >
                <Input
                  onChange={[Function]}
                  onClick={[Function]}
                />
                <SortBy
                  onChange={[Function]}
                  phoneNumbers={Array []}
                />
              </div>
              <ExportButton
                onClick={[Function]}
                phoneNumbers={Array []}
              />
              <GeneratedNumbers
                phoneNumbers={Array []}
              />
            </div>,
          ],
          "className": "wrapper",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {},
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <Error
                  error={false}
                  message=""
                />,
                <div
                  className="input-wrapper"
                >
                  <Input
                    onChange={[Function]}
                    onClick={[Function]}
                  />
                  <SortBy
                    onChange={[Function]}
                    phoneNumbers={Array []}
                  />
                </div>,
                <ExportButton
                  onClick={[Function]}
                  phoneNumbers={Array []}
                />,
                <GeneratedNumbers
                  phoneNumbers={Array []}
                />,
              ],
              "className": "App-body",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "error": false,
                  "message": "",
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": Array [
                    <Input
                      onChange={[Function]}
                      onClick={[Function]}
                    />,
                    <SortBy
                      onChange={[Function]}
                      phoneNumbers={Array []}
                    />,
                  ],
                  "className": "input-wrapper",
                },
                "ref": null,
                "rendered": Array [
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "function",
                    "props": Object {
                      "onChange": [Function],
                      "onClick": [Function],
                    },
                    "ref": null,
                    "rendered": null,
                    "type": [Function],
                  },
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "function",
                    "props": Object {
                      "onChange": [Function],
                      "phoneNumbers": Array [],
                    },
                    "ref": null,
                    "rendered": null,
                    "type": [Function],
                  },
                ],
                "type": "div",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "onClick": [Function],
                  "phoneNumbers": Array [],
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "phoneNumbers": Array [],
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
            ],
            "type": "div",
          },
        ],
        "type": "div",
      },
    ],
    "type": Symbol(react.fragment),
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <Header
            max={null}
            min={null}
            phoneNumbers={Array []}
            total={0}
          />,
          <div
            className="wrapper"
          >
            <SubHeader />
            <div
              className="App-body"
            >
              <Error
                error={false}
                message=""
              />
              <div
                className="input-wrapper"
              >
                <Input
                  onChange={[Function]}
                  onClick={[Function]}
                />
                <SortBy
                  onChange={[Function]}
                  phoneNumbers={Array []}
                />
              </div>
              <ExportButton
                onClick={[Function]}
                phoneNumbers={Array []}
              />
              <GeneratedNumbers
                phoneNumbers={Array []}
              />
            </div>
          </div>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "max": null,
            "min": null,
            "phoneNumbers": Array [],
            "total": 0,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <SubHeader />,
              <div
                className="App-body"
              >
                <Error
                  error={false}
                  message=""
                />
                <div
                  className="input-wrapper"
                >
                  <Input
                    onChange={[Function]}
                    onClick={[Function]}
                  />
                  <SortBy
                    onChange={[Function]}
                    phoneNumbers={Array []}
                  />
                </div>
                <ExportButton
                  onClick={[Function]}
                  phoneNumbers={Array []}
                />
                <GeneratedNumbers
                  phoneNumbers={Array []}
                />
              </div>,
            ],
            "className": "wrapper",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {},
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <Error
                    error={false}
                    message=""
                  />,
                  <div
                    className="input-wrapper"
                  >
                    <Input
                      onChange={[Function]}
                      onClick={[Function]}
                    />
                    <SortBy
                      onChange={[Function]}
                      phoneNumbers={Array []}
                    />
                  </div>,
                  <ExportButton
                    onClick={[Function]}
                    phoneNumbers={Array []}
                  />,
                  <GeneratedNumbers
                    phoneNumbers={Array []}
                  />,
                ],
                "className": "App-body",
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "function",
                  "props": Object {
                    "error": false,
                    "message": "",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": Array [
                      <Input
                        onChange={[Function]}
                        onClick={[Function]}
                      />,
                      <SortBy
                        onChange={[Function]}
                        phoneNumbers={Array []}
                      />,
                    ],
                    "className": "input-wrapper",
                  },
                  "ref": null,
                  "rendered": Array [
                    Object {
                      "instance": null,
                      "key": undefined,
                      "nodeType": "function",
                      "props": Object {
                        "onChange": [Function],
                        "onClick": [Function],
                      },
                      "ref": null,
                      "rendered": null,
                      "type": [Function],
                    },
                    Object {
                      "instance": null,
                      "key": undefined,
                      "nodeType": "function",
                      "props": Object {
                        "onChange": [Function],
                        "phoneNumbers": Array [],
                      },
                      "ref": null,
                      "rendered": null,
                      "type": [Function],
                    },
                  ],
                  "type": "div",
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "function",
                  "props": Object {
                    "onClick": [Function],
                    "phoneNumbers": Array [],
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "function",
                  "props": Object {
                    "phoneNumbers": Array [],
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
              ],
              "type": "div",
            },
          ],
          "type": "div",
        },
      ],
      "type": Symbol(react.fragment),
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
